### ISSUE TITLE
Server markes `assert_never` in exhaustive match statements as unnecessary

### ISSUE THREAD

**Issue by MrBhop** [astral-sh/ty#3312]
### Summary

After updating to `0.0.32` (from `0.0.31`) I noticed, that ty reports `assert_never` in exhaustive match statements as unreachable. Admittedly, ty is right about the code being unreachable, but i feel the diagnostic in this case is counterproductive. The intent of `assert_never` is to be unreachable and the diagnostic makes it seem like there is a mistake, when everything is working as expected.

Here's a reproducible example:
```python
from typing import assert_never
from enum import Enum


class Foo(Enum):
    BAR = "bar"
    BAZ = "baz"


def foo_func(foo: Foo):
    match foo:
        case Foo.BAR:
            ...
        case Foo.BAZ:
            ...
        case _:
            assert_never(foo) # ty: Code is unreachable  This may depend on your current environment and settings
```

Note, this only happens in my editor, not when running `ty check`.
If it's relevant I'm using ty in neovim via mason, without any custom config. 

### Version

ty 0.0.32

**Comment by sharkdp** [comment]
> Admittedly, ty is right about the code being unreachable, but i feel the diagnostic in this case is counterproductive. The intent of `assert_never` is to be unreachable and the diagnostic makes it seem like there is a mistake, when everything is working as expected.

Thank you for reporting this. We discussed this internally, and we do agree with you. The code is unreachable, but we shouldn't indicate to users that the code could/should be removed, because it does serve a purpose.

We eventually plan to add an (opt in) rule to ty that would allow you to catch non-exhaustive `match` statements *without* an `assert_never(…)` branch. So in that case, if you have that rule available, we feel like it's fine to show the code as grayed out, because in that case, the code is both unreachable and unnecessary.

**Comment by AlexWaygood** [comment]
> We eventually plan to add an (opt in) rule to ty that would allow you to catch non-exhaustive `match` statements _without_ an `assert_never(…)` branch. So in that case, if you have that rule available, we feel like it's fine to show the code as grayed out, because in that case, the code is both unreachable and unnecessary.

I would still personally want to include these branches as a matter of style, even if I opted into this rule. Some developers might prefer to have them removed, but I think these branches clearly signal to readers of the code that the branch is intended to be unreachable and the `match` is intended to be exhaustive, and readers of the code won't necessarily be reading the code with the assistance of a type checker or IDE (they may be reading it on GitHub, for example).

The sole purpose of `assert_never` is to declare to the type checker and to readers that the branch is intended to be deliberately unreachable. I think our default behaviour should be to never gray out these branches.

I would be okay with an _additional_ `unnecessary-assert-never` opt-in lint that can only be enabled if you also have the lint regarding non-exhaustive `match` statements enabled. But I think lots of users will want to retain these branches even if they have the lint regarding non-exhaustive `match` statements enabled.

**Comment by carljm** [comment]
It's not totally clear to me what the heuristic should be here. Do we skip the unreachable warning only if the entire suite consists only of an `assert_never` assertion?

**Comment by MichaReiser** [comment]
> It's not totally clear to me what the heuristic should be here. Do we skip the unreachable warning only if the entire suite consists only of an `assert_never` assertion?

That's what we had in mind. We could also only exclude the `assert_never` range (so that other statements in that branch are still marked as unreachable)

**Comment by AlexWaygood** [comment]
> Do we skip the unreachable warning only if the entire suite consists only of an `assert_never` assertion?

that's the heuristic that mypy uses when deciding when to avoid emitting a diagnostic warning that a block of code is accidentally unreachable

**Comment by MeGaGiGaGon** [comment]
One additional consideration that may come up is both [mypy](https://mypy-play.net/?mypy=latest&python=3.12&flags=strict%2Cwarn-unreachable&gist=edf6e1ca3b4889857cfe1ef0c2c1303c) and [basedpyright](https://basedpyright.com/?typeCheckingMode=all&strictParameterNoneValue=false&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgCwAUNdQCZHBQDGArgM4yQD6Kp5ACgCGINAC5C-EAEooAWgB8ksiDHUoGqCCFJ2RKABVERAKIhwIYaOm0aVBk2BgwAgB4TUMaWqqaoEIRhmAAsod3U-TWYhPUwUGAFvCMjIgDp05Mjo2O4fFJS2Th4%2BFTcbO2YAGxj2KABBPM0HFg4uCF4pAT1K4AAaKBFxZXJZRWHVTL9tXX0jBFNzXCs0cr9M5qcXbr6wj3ik33yAoNDww-yomP1PRMaLlPTUyfzs-Vzn%2B6ht1MK2jtLXNIgA) (though not normal [pyright](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgBQVAJkcFAMYCuAzjJAPoqnkAUAhiDQAuQjxABKKAFoAfGLIhhFKKqgh%2BSFkSgAVREQCiIcCAFCJVCrXrAwYXgA9RqGBOVqoEfjAYALKGcVTzUGfm1MFBhed2CQkIA6JLiQsIiOD3jU1nYILnEnSwoGABtwligAQUzVG0Yczm5FXm0S4AAaKEERBXIpOV6lFM8NLR19BCMTXHM0S08UursHVo7AlyjYrK8ff3XhrLSdVxia7cTk89DwnQyDq9WE5jZGgscJIA)) support "custom `assert_never`s" like this:
```py
from typing import Never

def custom_never(arg: Never) -> Never:
    raise TypeError(arg)

def foo(x: int):
    match x:
        case int():
            ...
        case _:
            custom_never(x)  # Not considered unreachable
```
I'm not sure exactly what the use case for it is, but ty may also want to support it.

**Comment by AlexWaygood** [comment]
Yes, I think the exact heuristic that mypy uses is something like "the block consists of only one statement, which is either a call that returns `Never` or a `raise` statement"

**Comment by sharkdp** [comment]
A similar case was discussed on Discord. We currently gray out `not TYPE_CHECKING` blocks, but they're obviously reachable at runtime:
```py
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    type MyType = Any
else:
    class MyType: ...  # grayed out in ty's LSP, but should certainly not be removed
```
https://play.ty.dev/409445e0-0337-437b-81d7-4d419a24495b

**Comment by AlexWaygood** [comment]
> A similar case was discussed on Discord. We currently gray out `not TYPE_CHECKING` blocks, but they're obviously reachable at runtime

The list of heuristics we need to apply here to avoid greying out "deliberately unreachable" code starts to look very similar to the list of heuristics I described in #1948 for avoiding emitting diagnostics on deliberately unreachable code ;)

**Comment by carljm** [comment]
I originally thought it would be necessary to ensure we don't ever gray out code that can run at runtime (interpreting the graying out as "you can remove this code" signal.) But we've since discussed that this is impossible. A condition that originates from `sys.platform` or `sys.version_info` can be laundered through other names, and "choice of venv" is also a configuration option that can change whether some code is reachable or not, and we can't possibly detect that.

For comparison, rust analyzer makes no attempt to do this at all; even explicitly configuration gated code sections are still grayed out and give me a diagnostic in nvim. 

We could say "ok it's impossible to fully do this, but we can still attempt to do it in the obvious cases and just accept we'll still have false positives in less common cases." (I'm not sure how uncommon the less common cases are; it'd be easy to eg a hasattr check on a typeshed type and make an effectively version dependent reachability condition that we would not be able to detect.)

Or we could take the rust analyzer route and say, no, graying out is not "you should delete this code", it's simply informing you that "the type checker is considering this code unreachable under the current checking config." This does seem like useful information. 

Have we looked at what other LSPs do here?

**Comment by MichaReiser** [comment]
> Have we looked at what other LSPs do here?

At least for `assert_never`, Pylance and Pyrefly don't gray out the code

**Comment by carljm** [comment]
Oh my comment really should have gone on #1948 or a more general issue. I agree we should probably suppress for blocks containing only `assert_never` or a `raise`. That we can do reliably, it's always local, and indicates the user is already aware that code is unreachable. More thinking about `if not TYPE_CHECKING` and the config dependent cases.

**Comment by AlexWaygood** [comment]
> I originally thought it would be necessary to ensure we don't ever gray out code that can run at runtime (interpreting the graying out as "you can remove this code" signal.) But we've since discussed that this is impossible. A condition that originates from sys.platform or sys.version_info can be laundered through other names, and "choice of venv" is also a configuration option that can change whether some code is reachable or not, and we can't possibly detect that.

You're correct that doing this fully rigorously with 100% accuracy is, of course, impossible in our model that uses a much more generalised, type-aware approach to control flow than other type checkers. I don't think that contradicts any of the following points, however:
1. All other type checkers offer (either opt-in or opt-out) diagnostics that fire if they identify a block of code that looks like it is accidentally unreachable. This usually indicates a bug in the user's code; users find these diagnostics very useful
2. All other type checkers try to suppress these diagnostics for blocks that only consist of `assert_never()` calls or `raise` statements, blocks that are unreachable due to `if TYPE_CHECKING` checks, blocks that are unreachable to due `sys.platform` checks, or blocks that are unreachable due to `sys.version_info` checks
4. We should aim to achieve feature parity with other type checkers when it comes to features that users find very useful
5. In order to implement our version of this diagnostic, we will also need to find a way to suppress unreachable-code diagnostics for various blocks that appear to be deliberately unreachable, similar to (2). Our version of this diagnostic will simply have too many noisy false positives if we do otherwise.
6. Just because we cannot identify `sys.version_info` blocks with 100% accuracy, for the reasons you identify, does not mean that we cannot attempt some heuristics that will lead to us avoiding emitting unreachable-code diagnostics for the vast majority of `sys.version_info` branches (for example)
7. Once you have conceded that we need to implement these heuristics anyway in order to achieve feature parity with other type checkers, we may as well apply the same heuristics for graying out unreachable code as we do for emitting diagnostics on accidentally unreachable code. This seems like the most consistent and simplest course of action

**Comment by carljm** [comment]
Makes sense. I do worry that our more sophisticated ability to detect unreachable code, combined with our inability to always detect when that unreachability is intended/conditional, will mean we have more trouble than other type checkers with "false positives" if we emit a more aggressive unreachable code diagnostic. And I'm not sure what solutions we'd have available for that problem. We could artificially restrict our unreachability detection to match what other type checkers can do?

(Edit: I guess in effect what I'm saying here is that I'm not sure even best-effort suppressing actually gives us "feature parity" with other type checkers, unless we also hamstring our reachability detection to match theirs.)

But we can try it and see how it goes in practice.

**Comment by carljm** [comment]
It does seem plausible to me that our best option for feature parity and avoiding false positives in #1948 will be to only emit the diagnostic on code we are able to identify as unreachable in semantic indexing, and not use type aware reachability. And at that point it becomes less clear to me how much parity we should aim for between our aggressive unreachable diagnostic and our greying-out info.

**Comment by AlexWaygood** [comment]
> It does seem plausible to me that our best option for feature parity and avoiding false positives in [#1948](https://github.com/astral-sh/ty/issues/1948) will be to only emit the diagnostic on code we are able to identify as unreachable in semantic indexing, and not use type aware reachability. And at that point it becomes less clear to me how much parity we should aim for between our aggressive unreachable diagnostic and our greying-out info.

The way this issue was reported by the OP (a neovim user) indicates that we are really already issuing disruptive diagnostics if you're using a non-VSCode editor. While VSCode filters out "hint"/"info"-level diagnostics from the "problems" panel, it appears that other editors do not do similarly. So I still don't see a good case for applying different heuristics when it comes to greying out code (implemented as a "hint"-level diagnostic) versus a warning/error-level diagnostic.

For neovim users, it really seems like we've already accidentally implemented #1948, but without any of the heuristics proposed in that issue to avoid false positives.

**Comment by carljm** [comment]
I think that's correct -- but also at some point has to be on the editors / users? I'm sure it's possible to configure nvim to not show hint/info in problems pane. There's not much point in severity levels if we have to treat them all the same.

**Comment by AlexWaygood** [comment]
Sure, but it's obviously better for users if ty gives a great experience out of the box without any special configuration on their part. It still seems to me that any heuristic or principle we apply to avoid false positives for warning/error-level diagnostics regarding accidentally unreachable code should also probably be applied for the diagnostics we already emit to achieve dimming of unreachable code. False positives for the info-level diagnostics are already causing issues for neovim users; it seems like these false positives are just as problematic as false positives for any other diagnostic.

**Comment by AlexWaygood** [comment]
FWIW, I think it actually would be quite easy to rigorously track which boolean conditions in control flow arose from `sys.version_info` checks. You're right that the symbol `sys.version_info` can be arbitrarily aliased, but `sys.version_info` is an instance of the tuple subclass `sys._version_info`, which is a singleton type. We could easily add special casing so that any comparison involving an instance of `sys._version_info` specifically evaluates to a "tagged boolean-literal type" that marks it as having originated from a `sys.version_info` check. That information could then be used to inform whether to emit unreachable code diagnostics.

`typing.TYPE_CHECKING`, `sys.platform` and `os.name` chekcs would be harder, however.

**Comment by carljm** [comment]
> FWIW, I think it actually would be quite easy to rigorously track which boolean conditions in control flow arose from `sys.version_info` checks. You're right that the symbol `sys.version_info` can be arbitrarily aliased, but `sys.version_info` is an instance of the tuple subclass `sys._version_info`, which is a singleton type. We could easily add special casing so that any comparison involving an instance of `sys._version_info` specifically evaluates to a "tagged boolean-literal type" that marks it as having originated from a `sys.version_info` check. That information could then be used to inform whether to emit unreachable code diagnostics.

This approach would only handle the very specific case of direct aliasing. It wouldn't even handle something like this:

```py
if sys.version_info >= (3, 11):
    NEW_PYTHON = True
else:
    NEW_PYTHON = False
```

or any more complex variants. It doesn't seem likely to me that it would move the needle enough on real-world false positives to be worth the implementation.

> it seems like these false positives are just as problematic as false positives for any other diagnostic.

But "info" or "hint" diagnostics are not claiming to represent an actionable problem, so I wouldn't even call these false positives. As a user, I'd find it useful (and I do find it useful when rust-analyzer does it) to gray out all code that my type checker considers unreachable under its current configuration; it would help me to understand its other behaviors (e.g. why is this definition of a variable not respected in its later type, etc.)
